Re-work [Java]

ZimCodes opened the pull request:
### Summary
<!-- Provide a general summary of your pull request here -->
Zyod will be re-written to Java.

#### Detailed Info
<!-- Describe the pull request in greater detail-->
Zyod is currently written in Python 3.10 *as of this pr*. 

At the moment, in order to use Zyod, all project file must be downloaded. Although the project could utilize other tools to bundle the files into a single executable, I prefer to use more native options. Bundling an entire project into a `.jar` file is more appropriate in this manner. Moving forward, Zyod would benefit from the type safety features Java offers, providing a well-structured code base. 

#### References
<!-- Are there any references you can provide which motivated these changes? -->
<!-- If it fixes an open issue, please link to the issue here. -->
None

#### Screenshots (if appropriate):

ZimCodes committed:
Update gitignore file;
Remove Python & add the base Java project for Zyod;

ZimCodes committed:
Adds properties file;
Implement arguments & tests for general and web driver;

ZimCodes committed:
All options are now available;
More tests for the command line arguments are added;
Update README with new information;
lombok dependency;
Reduce scope modifiers of argument parameters from public to private;

ZimCodes committed:
Make it easier to retrieve properties for the app;
Increase app version to be ahead of python's version;
Removes option `--driver-path` which will manually look for a webdriver executable and instead opt-in to use an automatic method. Add WebDriverManager dependency;
Adds new `--all-certs` & `--compat-driver` options;
Refactor test code & fixtures;
Removes Lombok dependency and implement own methods;
Reduce modifier scopes of classes & methods;
More custom assertion for tests;
Implement webdriver classes for edge,firefox,& chrome browsers;
Update all build dependencies;
Update README.md;
Add WebDriver fixtures;
Adds a resource file containing variety of different mime types;
Add tests for webdriver settings;
Adds Writer & Talker classes. Writer class handles reading/writing operation for files. Talker writes messages to the command line;

ZimCodes committed:
Refactor argument related content including tests, fixtures, and classes;
Combine all arguments into a singular Arg class;
Update driver tests with new argument changes;
Add new driver test for firefox to detect headless settings separately;
Update Talker class;
Update Writer class with the ability to read URLs from an input file;

ZimCodes committed:
Refactor fixtures;
Rename browser name from 'edge' to 'msedge';
Adds new junit dependency for ParameterizedTests;
Rename package name 'drivers' to 'browsers';
Removes Driver fixture in favor of a global one;
Update driver setting tests;
Add BrowserType enums for fixed browser names;
Create a Driver Factory class;
Adds new Driver factory test;

ZimCodes committed:
Finish setting up navigation information & OD metadata;
Creates a mutable URL class;
Adds testing for URL class;

ZimCodes committed:
New app property for download directory;
Set default new download directory to `Download folder/Zyod`;
Refactor app configurations to contain resources;
Move line separator from GlobalDefault to AppConfig;
For wait options, use milliseconds instead of seconds. This means changing the type from an `int` to a `long`. Also a millisecond converter is also added;
Update tests & custom asserts;
New dependency `GSON`. This is used to parse JSON files;
Set implicit wait options for all browser types;
NavInfo now holds the OD name & Navtype;
NavInfo now overrides the `hashcode()` & `equals()` methods to be used in a map;
Set defaults for NavInfo Builder;
Update README;
Refactor custom asserts from a single file to multiple files depending on test;
Adds pr template;
Adds interactive support functions for browsers;
Place navigation information inside a JSON file;
Adds support class for navigation;
Store references to demo ODs;
Adds default download directory preference to chromium browsers;

ZimCodes committed:
Add simulated interactivity tests. Exclude said test by default;
Remove empty class;
Add new navigation recipe info to JSON, `download_filter` to filter out downloadble contents. Rename navigation info `reject_filter` to `file_reject_filter`;
Move NavInfoParser to the same package as Navinfo;
Update not elements error message;
Update Talker util class;
Adds new asset Directory & Downloader functionality;

ZimCodes committed:
Relocate all CSS paths for download, navigation, & scraping to individual JSON files. Another JSON has been added, `download_recipe.json`, to hold all CSS pathing data for downloading;
Adds the locations of JSON resources as a configurable in app.properties file;
Path of test resource folder is now in GlobalDefault;
Adds new DownloadInfo which holds information about downloading;
Adds single & multi-download variations of right-click downloads & regular downloads;
Switch DownloadFilter to extend BiFunction instead of Function interface;
Rename `NavInfo` test to just `InfoTest` as it now tests all CSS metadata from multiple JSON files;
Removes one GoIndex reference opting for just one reference per OD;
Adds new css navigation paths, going back to a previous page or home page;
Implement download filters;
Update tests for downloading;

ZimCodes committed:
Implements the identifier, which will identify the type of OD currently present;
Update tests for identifier;

blank-manash commented:
Hi, I read about your project and I'm interested in helping you out. I'm have experience with Java and can resolve the rework if you are interested. 

However, I might need a little bit more context on how this projects works and is implemented. If that's alright, do let me know.

ZimCodes committed:
Update download & navigation information;

ZimCodes commented:
> Hi, I read about your project and I'm interested in helping you out. I'm have experience with Java and can resolve the rework if you are interested.
>
>However, I might need a little bit more context on how this projects works and is implemented. If that's alright, do let me know.

## Work In Progress
@blank-manash Thank you for the offer. Part of the reason why I wanted to rework this project is to provide an unclouded code structure as well as to learn Java myself. There are multiple components Zyod has in which I plan to add, subtract, and modify away from the previous Python edition. Essentially, this pr will establish a new base foundation for Zyod.

For now, I would like to take the time to unravel Zyod's potential going forward. 

Until I create a guideline for contributors providing thorough context on Zyod and how it works, I will not be accepting prs at the moment.

## Contributing

You can still contribute by finding Dynamic ODs *(Open Directories)* Zyod currently does not support. Specifically, ODs that will only load if JavaScript is enabled. **Be sure to create an issue for it**.

Here is an example of the [GDIndex demo OD](https://gdindex-demo.maple3142.workers.dev). Try disabling JavaScript, refresh the page, and you will see a blank website. It is as if it never existed.

Additionally, I created a [CONTRIBUTING](https://github.com/ZimCodes/Zyod/blob/main/CONTRIBUTING.md) file, providing a bit more info about some key components of Zyod that will be added.

ZimCodes committed:
update .gitignore;
Moves `--wait` & `--random-wait` arguments to the new Scraper argument category;
Removes `--web-wait` option;
Exclude AppTest from tests;
update Directory class;
Place keys of map in download filter as a singular field;
Improve ODURL asset. When an illegal character exception is thrown the link will now be URL encoded;
 ODURL `joiner()` now returns an ODURL instead of a String;
 Adds comparator methods to compare two URLs regardless if they are encoded;
 Update README;
 Change default of `--wait` option to 4 seconds;
 Update tests;
 Implement all ScrapeFilters;
 Adds a Scraper called JoinScraper;
 Adds tests for ScrapeFilters and JoinScraper;
 Adds a new test to compare ODURLs;

ZimCodes committed:
Fix some characters being seen as illegal under the windows-31j encoding;
Rename `--load-wait` option to `--page-wait`;
New option `--element-wait`. Implicitly wait for web elements to make an appearance;
update tests & fixtures;
Rename `JoinScraper` to `AttributeScraper`;
More "NOTE:"s;
Adds waiting before right-clicking to download. The waiting is also controlled by `--ddwait`;
 Update query regex to remove URL queries that has characters other than the latin alphabet. This regex also removes URL fragments as a side effect;
 Update nav_recipes metadata for AList ODs;
 Makes the retrieval of a single element more efficient when using NavSupport util class;
 Update File Regex. This regex determines if text is a file or not. In addition, the isFile() has been improved and now utilizes java's URL class to help in determining if a piece of text is a file or not;
 Update README;
Close drivers after usage in tests;
Adds TextScraper & SiftScraper;

ZimCodes committed:
Rename `ODID` to `OdIdentifier`. This is the new name for the base class of OD identifiers;
Implement a separate WebElement retrieval method for OD identifiers to use;
Change all hanzi characters to their unicode counterpart. This gets rid of the compiler replacing some hanzi characters with a `�`;
Remove unused tests;
Repurpose refresh option. `--refresh` is now `--no-refresh`. Refreshing is enabled by default now;
New option `--interact-wait`. This option configures the amount of seconds to wait before/after a simulated interaction (clicks,scrolls,etc);
Update tests & assertions;
Rename base class `ODScraper` to `Scraper`;
Fix creating a new directory after joining link when the maximum depth level is reached for scraping;
Rename `NavigatorFactory` to be `BrowserFactory`. BrowserFactory creates new `Browser` objects which will in turn allows for the creation of its associated RemoteWebDriver;
Browser class will now create a separate driver for OD identifiers to use. These drivers are faster than the regular drivers as wait periods are limited;
Fix possible chance a Directory object may not have a URL;
Fix some DLFilers not working properly;
Remove download recipe for GoIndex thumbnail variant;
Add more filter recipes for downloading in download_recipes.json;
Fix single download method not filtering out unnecessary resources;
Rename `ODDefault` to `ODDemoRef` for tests;
Create a new functionalInterface for extra downloading tasks, `DownloadTasks`;
Update regex for query to include queries without a value, but has a key;
Update scrape/nav recipes for various ODs;
Adds a new navigation type for ShareList called `List`. Resources setup in a list for this navigation variant;
Implement NavigatorFactory. The factory creates the appropriate navigator to the current OD. To determine this, ODIdentifiers are used to find unique elements to identify the ODType;
`ODURL` now removes square brackets upon initialization;
Update README;
Move assertions from `URLDefault` to `URLAssert`;
Adds all navigators;
Adds many more tests;

ZimCodes committed:
Clean up custom scraper test;
Update Talker class with more text to write to the console;
Adds documentation to Writer class;

ZimCodes committed:
Zyod is now put together;
Update tests;
Add 2 new options: `--init-page-wait` & `--init-refresh`;
`--wait` default is now `6` seconds;
Update gradle dependencies;
Update hashcode & equals for Directory class;
Allow multiple remotewebdrivers to be used when identifying the correct Navigator type to use for the current url;
Update OD references for tests;
Update README.md file. This includes adding a new section, *Things to Note*;
Fix starting cursor of touch navigator pointing to the wrong URL;
Finished Writer class. Add the ability to record URLs found in the OD and write this info to a file;
Adds a new test for the Writer class;
Adds a new class, Zyod, which combines & utilizes all components of Zyod;
Display & Terminate app version  if `--version` option is specified;

ZimCodes committed:
Adds a new test for command line arguments;
Update downloading process & recipes for ZFile ODs;
Removes ZFile OD's download filter;
Update tests to match new downloading process for ZFile ODs;
Adds a new browser configuration for Firefox. `media.play-stand-alone` prevents video URLs from playing the file in the browser. This will force the browser to download the file instead. So there is no configuration like this for Chromium browsers;
Update tests;
Update README.md;

ZimCodes committed:
Update README.md

ZimCodes committed:
Update README

ZimCodes committed:
Merge branch 'main' of https://github.com/ZimCodes/Zyod

ZimCodes committed:
merging into main
